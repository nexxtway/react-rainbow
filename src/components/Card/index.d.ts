import { ReactNode } from 'react';
import { BaseProps } from '../types';

export interface CardProps extends BaseProps {
    title?: ReactNode;
    icon?: ReactNode;
    actions?: ReactNode;
    footer?: ReactNode;
    children?: ReactNode;
    isLoading?: boolean;
}

export default function(props: CardProps): JSX.Element | null;
