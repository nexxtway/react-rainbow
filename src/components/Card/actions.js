import React from 'react';
import PropTypes from 'prop-types';
import './styles.css';

export default function Actions({ content }) {
    if (content) {
        return <div className="rainbow-no-flex">{content}</div>;
    }
    return null;
}

Actions.propTypes = {
    content: PropTypes.node,
};
Actions.defaultProps = {
    content: null,
};
