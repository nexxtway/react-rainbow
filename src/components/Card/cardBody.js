import React from 'react';
import PropTypes from 'prop-types';
import Spinner from '../Spinner';

export default function CardBoddy({ isLoading, children }) {
    if (isLoading) {
        return <Spinner isVisible={isLoading} size="small" />;
    }
    return children;
}

CardBoddy.propTypes = {
    isLoading: PropTypes.bool,
    children: PropTypes.node,
};
CardBoddy.defaultProps = {
    isLoading: null,
    children: null,
};
