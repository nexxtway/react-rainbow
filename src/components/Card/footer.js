import React from 'react';
import PropTypes from 'prop-types';
import './styles.css';

export default function Footer({ content, isLoading }) {
    if (content && !isLoading) {
        return (
            <footer className="rainbow-card__footer">
                {content}
            </footer>
        );
    }
    return null;
}

Footer.propTypes = {
    content: PropTypes.node,
    isLoading: PropTypes.bool,
};

Footer.defaultProps = {
    content: null,
    isLoading: null,
};
