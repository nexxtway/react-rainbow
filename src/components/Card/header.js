import React from 'react';
import PropTypes from 'prop-types';
import HeaderTitle from './headerTitle';
import RenderIf from '../RenderIf';
import StyledHeaderContainer from './styled/headerContainer';
import StyledHeader from './styled/header';
import StyledIcon from './styled/icon';

export default function Header({ icon, title, actions }) {
    return (
        <RenderIf isTrue={!!(icon || title || actions)}>
            <StyledHeaderContainer>
                <StyledHeader>
                    <RenderIf isTrue={!!icon}>
                        <StyledIcon>{icon}</StyledIcon>
                    </RenderIf>
                    <HeaderTitle title={title} />
                </StyledHeader>
                <RenderIf isTrue={!!actions}>
                    <div>{actions}</div>
                </RenderIf>
            </StyledHeaderContainer>
        </RenderIf>
    );
}

Header.propTypes = {
    title: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    icon: PropTypes.node,
    actions: PropTypes.node,
};

Header.defaultProps = {
    title: null,
    icon: null,
    actions: null,
};
