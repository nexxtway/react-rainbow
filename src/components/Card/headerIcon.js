import React from 'react';
import PropTypes from 'prop-types';
import './styles.css';

export default function HeaderIcon({ icon }) {
    if (icon) {
        return (
            <div className="rainbow-card-media__figure">
                {icon}
            </div>
        );
    }
    return null;
}

HeaderIcon.propTypes = {
    icon: PropTypes.node,
};
HeaderIcon.defaultProps = {
    icon: null,
};
