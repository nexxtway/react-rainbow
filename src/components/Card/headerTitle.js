/* eslint-disable no-script-url */
import React from 'react';
import PropTypes from 'prop-types';
import './styles.css';

export default function HeaderTitle({ title }) {
    if (typeof title === 'string') {
        return (
            <h2>
                <a href="javascript:void(0);" className="rainbow-card_header-link">
                    <span className="rainbow-card_title">{title}</span>
                </a>
            </h2>
        );
    }
    return title;
}

HeaderTitle.propTypes = {
    title: PropTypes.string,
};
HeaderTitle.defaultProps = {
    title: null,
};
