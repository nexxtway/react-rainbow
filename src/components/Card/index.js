import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import CardBoddy from './cardBody';
import Header from './header';
import RenderIf from '../RenderIf';
import './styles.css';

/**
* Cards are used to apply a container around a
* related grouping of information.
*/
export default function Card(props) {
    const {
        className,
        style,
        actions,
        children,
        footer,
        title,
        icon,
        isLoading,
    } = props;

    const getContainerClassName = () => classnames('rainbow-card', className);

    return (
        <article className={getContainerClassName()} style={style}>
            <Header actions={actions} title={title} icon={icon} />
            <div>
                <CardBoddy isLoading={isLoading}>
                    {children}
                </CardBoddy>
            </div>

            <RenderIf isTrue={footer && !isLoading}>
                <footer className="rainbow-card_footer">
                    {footer}
                </footer>
            </RenderIf>
        </article>
    );
}

Card.propTypes = {
    /** The title can include text or another component,
    * and is displayed in the header of the component. */
    title: PropTypes.oneOfType([
        PropTypes.string, PropTypes.node,
    ]).isRequired,
    /** The icon to show if it is passed. It must be a svg icon or a font icon.
    * It is displayed in the header of the component. */
    icon: PropTypes.node,
    /** Actions are components such as button or buttonIcon. Actions are displayed in the header. */
    actions: PropTypes.node,
    /** The footer can include text or another component. */
    footer: PropTypes.oneOfType([
        PropTypes.string, PropTypes.node,
    ]),
    /** The content of the Card. */
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.object,
    ]),
    /** If is set to true, then is showed a loading symbol. */
    isLoading: PropTypes.bool,
    /** A CSS class for the outer element, in addition to the component's base classes. */
    className: PropTypes.string,
    /** An object with custom style applied to the outer element. */
    style: PropTypes.object,
};

Card.defaultProps = {
    icon: null,
    actions: null,
    footer: null,
    isLoading: false,
    className: undefined,
    style: undefined,
    children: null,
};
