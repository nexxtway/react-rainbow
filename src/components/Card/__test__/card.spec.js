import React from 'react';
import { shallow } from 'enzyme';
import Card from './../index';

describe('<Card/>', () => {
    it('should set to true the isTrue prop in RenderIf', () => {
        const component = shallow(
            <Card
                title="Card Header"
                footer="Card Footer"
                actions={<div>Testing actions node</div>}
            >
                Testing childrens
            </Card>,
        );
        expect(component.find('RenderIf').prop('isTrue')).toBe(true);
    });
    it('should set to true the isTrue prop in RenderIf when isLoading', () => {
        const component = shallow(
            <Card
                title="Card Header"
                footer="Card Footer"
                isLoading
                actions={<div>Testing actions node</div>}
            >
                Testing childrens
            </Card>,
        );
        expect(component.find('RenderIf').prop('isTrue')).toBe(false);
    });
    it('should set to true the isTrue prop in RenderIf when footer is not passed', () => {
        const component = shallow(
            <Card title="Card Header" actions={<div>Testing actions node</div>}>
                Testing childrens
            </Card>,
        );
        expect(component.find('RenderIf').prop('isTrue')).toBe(false);
    });
});
