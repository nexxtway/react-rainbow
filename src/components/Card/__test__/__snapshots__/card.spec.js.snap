// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card/> should have the right class names 1`] = `
<article
  className="rainbow-card"
>
  <div
    className="rainbow-card_header-container"
  >
    <header
      className="rainbow-card_header"
    >
      <h2>
        <a
          className="rainbow-card_header-link"
          href="javascript:void(0);"
        >
          <span
            className="rainbow-card_title"
          >
            Card Header
          </span>
        </a>
      </h2>
    </header>
    <div>
      <div>
        Testing actions node
      </div>
    </div>
  </div>
  <div>
    <div>
      Testing childrens
    </div>
  </div>
  <footer
    className="rainbow-card_footer"
  >
    Card Footer
  </footer>
</article>
`;
